mcf-browser/src/back-end/parse.ts

```diff
@@ -0,0 +1,185 @@
+  * @return {Object} an object containing all of the ids of the
+  *    subject nodes and the error messages
+  */
+async function getNodes(fileList: Blob[]) {
+  const finalReturn: ParseFileResponse = {
+    errMsgs: [],
+    localNodes: [],
+    datapoints: [],
+  };
+
+  // Find TMCF file, if it exists
+  let tmcfFile = null;
+  for (const file of fileList) {
+    const fileName = (file as File).name;
+    const fileExt = fileName.split('.').pop();
+
+    if (fileExt === 'tmcf') {
+      if (tmcfFile) {
+        // If another TMCF file was found, throw an error
+        finalReturn['errMsgs'] = finalReturn['errMsgs'].concat([{
+          'file': (tmcfFile as File).name,
+          'errs': [
+            ['-1', '', ERROR_MESSAGES.MULTIPLE_TMCF],
+          ],
+        }]);
+      }
+
+      tmcfFile = file;
+    }
+  }
+
+
+  for (const file of fileList) {
+    const fileName = (file as File).name;
+    const fileExt = fileName.split('.').pop();
+
+    if (fileExt === 'mcf') {
+      const mcfOut = await ParseMcf.readFile(file);
+
+      if (mcfOut['errMsgs'].length !== 0) {
+        finalReturn['errMsgs'] = finalReturn['errMsgs'].concat([
+          {
+            file: fileName,
+            errs: mcfOut['errMsgs'],
+          },
+        ]);
+      }
+
+      finalReturn['localNodes'] = finalReturn['localNodes'].concat(
+          mcfOut['localNodes'],
+      );
+    } else if (fileExt === 'csv') {
+      if (tmcfFile) {
+        const tmcfFileName = (tmcfFile as File).name;
+        const tmcfOut = await ParseTmcf.generateMcf(tmcfFile, file).then(
+            (mcf) => {
+              const mcfParser = new ParseMcf(
+                  tmcfFileName + '&' + fileName,
+              );
+              return mcfParser.parseMcfStr(mcf as string);
+            },
+        );
+
+        const datapoints = await ParseTmcf.generateDataPoints(tmcfFile, file);
+        finalReturn['datapoints'] = finalReturn['datapoints'].concat([
+          datapoints,
+        ]);
+
+        if (tmcfOut['errMsgs'].length !== 0) {
+          finalReturn['errMsgs'] = finalReturn['errMsgs'].concat({
+            file: tmcfFileName,
+            errs: tmcfOut['errMsgs'],
+          });
+        }
+        finalReturn['localNodes'] = finalReturn['localNodes'].concat(
+            tmcfOut['localNodes'],
+        );
+      }
+    }
+  }
+  return finalReturn;
+}
+
+/** Group nodes and find all time series
+  * @param {Object[]} datapoints the time series data
+  * @return {Series[]} an array of time series in the data
+  */
+function getTimeData(datapoints: Object[]) {
+  // Turn from array of objects (one per file) to one big object
+
+  const allData: any = {};
+  for (const data of datapoints) {
+    const allSeries = Object.keys(data);
+    for (const series of allSeries) {
+      allData[series] = allData[series] ? allData[series] : {};
+      allData[series] = {...allData[series], ...(data as any)[series]};
```

Comment:
don't you want to add `, ... allSeries[series]};` ?

Reply:
allSeries (defined on line 131) is a list of the keys for data so allSeries[series] would throw an error (since it's a list not an object).

Reply:
oh, I see, this was confusing because types weren't specified. Please replace `any` type with the actual type and `Object` with that actual type. Please do this for here and everywhere else in this PR.

Also, I don't think (data as any) is needed?

Reply:
Made all of the changes in the next commit! Thanks for the feedback